File: guild_of_heroes/guild_of_heroes_tests/guild_tests.cpp
#include "pch.h"
#include "guild_of_heroes.hpp"

using namespace guild_of_heroes;

TEST(GuildTests, create_creationWithSomeName_creationOk) {
	Guild* guild = Guild::create("my guild");
	EXPECT_NE(guild, nullptr);
	delete guild;
}
TEST(GuildTests, create_creationWithEmptyName_throwsException) {
	EXPECT_THROW_WITH_MESSAGE(
		Guild::create(""), 
		std::invalid_argument,
		"name can not be empty"
	);
}
TEST(GuildTests, create_creationWithSomeName_heroesCountZero) {
	Guild* guild = Guild::create("my guild");
	EXPECT_EQ(guild->getHeroesCount(), 0);
	delete guild;
}
TEST(GuildTests, create_creationWithSomeName_heroesListIsEmpty) {
	Guild* guild = Guild::create("my guild");
	EXPECT_EQ(guild->getHeroes(), std::vector<Hero*>());
	delete guild;
}

TEST(GuildTests, getName_creationWithSomeName_returnCorrectName) {
	Guild* guild = Guild::create("my guild");
	EXPECT_EQ(guild->getName(), "my guild");
	delete guild;
}

TEST(GuildTests, addHero_addFirstHero_heroesCountOne) {
	Guild* guild = Guild::create("my guild");
	Hero* hero = Hero::create("my hero");
	guild->addHero(hero);
	EXPECT_EQ(guild->getHeroesCount(), 1);
	delete guild;
	delete hero;
}
TEST(GuildTests, addHero_addSecondHero_heroesCountIncrement) {
	Guild* guild = Guild::create("my guild");
	Hero* hero1 = Hero::create("my hero 1");
	Hero* hero2 = Hero::create("my hero 2");
	guild->addHero(hero1);
	EXPECT_EQ(guild->getHeroesCount(), 1);
	guild->addHero(hero2);
	EXPECT_EQ(guild->getHeroesCount(), 2);
	delete guild;
	delete hero1;
	delete hero2;
}
TEST(GuildTests, addHero_addNullHero_throwsException) {
	Guild* guild = Guild::create("my guild");	
	EXPECT_THROW_WITH_MESSAGE(
		guild->addHero(nullptr),
		std::invalid_argument,
		"hero can not be null"
	);
	delete guild;
}

TEST(GuildTests, removeHero_removeExistHeroHero_heroesCountDecrement) {
	Guild* guild = Guild::create("my guild");
	Hero* hero1 = Hero::create("my hero 1");
	Hero* hero2 = Hero::create("my hero 2");
	guild->addHero(hero1);
	guild->addHero(hero2);
	EXPECT_EQ(guild->getHeroesCount(), 2);
	guild->removeHero(hero1);
	EXPECT_EQ(guild->getHeroesCount(), 1);
	delete guild;
	delete hero1;
	delete hero2;
}

File: GuildOfHeroes/GuildOfHeroes.Tests/GuildTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using GuildOfHeroes.Core;
using NUnit.Framework;

namespace GuildOfHeroes.Tests
{
    class GuildTests
    {
        [Test]
        public void Create_creationWithSomeName_creationOk()
        {
            Guild guild = Guild.Create("my guild");
            Assert.IsNotNull(guild);            
        }
		[Test]
		public void Create_creationWithEmptyName_throwsException()
		{
			var exc = Assert.Throws<ArgumentException>(
				() => Guild.Create("")
			);
			Assert.IsTrue(
				exc.Message.ToLower().Contains("name can not be empty")
			);
		}
		[Test]
        public void Create_creationWithSomeName_heroesCountZero()
		{
			Guild guild = Guild.Create("my guild");
			Assert.AreEqual(0, guild.HeroesCount);
		}
		[Test]
        public void Create_creationWithSomeName_heroesListIsEmpty()
		{
			Guild guild = Guild.Create("my guild");
			var heroes = guild.Heroes;
			Assert.AreEqual(0, heroes.Count);
		}

		[Test]
        public void GetName_creationWithSomeName_returnCorrectName()
		{
			Guild guild = Guild.Create("my guild");
			Assert.AreEqual("my guild", guild.Name);
		}

		[Test]
        public void AddHero_addFirstHero_heroesCountOne()
		{
			Guild guild = Guild.Create("my guild");
			Hero hero = Hero.Create("my hero");
			guild.AddHero(hero);
			Assert.AreEqual(1, guild.HeroesCount);
		}
		[Test]
        public void AddHero_addSecondHero_heroesCountIncrement()
		{
			Guild guild = Guild.Create("my guild");
			Hero hero1 = Hero.Create("my hero 1");
			Hero hero2 = Hero.Create("my hero 2");
			guild.AddHero(hero1);
			Assert.AreEqual(1, guild.HeroesCount);
			guild.AddHero(hero2);
			Assert.AreEqual(2, guild.HeroesCount);
		}
		[Test]
        public void AddHero_addNullHero_throwsException()
		{
			Guild guild = Guild.Create("my guild");
			var exc = Assert.Throws<ArgumentException>(
				() => guild.AddHero(null)
			);
			Assert.IsTrue(
				exc.Message.ToLower().Contains("hero can not be null")
			);
		}
		[Test]
        public void AddHero_addExistHero_throwsException()
		{
			Guild guild = Guild.Create("my guild");
			Hero hero1 = Hero.Create("my hero 1");
			guild.AddHero(hero1);
			var exc = Assert.Throws<ArgumentException>(
				() => guild.AddHero(hero1)
			);
			Assert.IsTrue(
				exc.Message.ToLower().Contains("hero already added")
			);
		}

		[Test]
		public void GetHeroes_addTwoHeroes_resultContainsBoth()
		{
			Guild guild = Guild.Create("my guild");
			Hero hero1 = Hero.Create("my hero 1");
			Hero hero2 = Hero.Create("my hero 2");
			guild.AddHero(hero1);
			guild.AddHero(hero2);

			var heroes = guild.Heroes;
			Assert.IsTrue(heroes.Contains(hero1));
			Assert.IsTrue(heroes.Contains(hero2));
			Assert.AreEqual(2, heroes.Count);
		}

		[Test]
        public void RemoveHero_removeExistHero_heroesCountDecrement()
		{
			Guild guild = Guild.Create("my guild");
			Hero hero1 = Hero.Create("my hero 1");
			Hero hero2 = Hero.Create("my hero 2");
			guild.AddHero(hero1);
			guild.AddHero(hero2);
			Assert.AreEqual(2, guild.HeroesCount);
			guild. RemoveHero(hero1);
			Assert.AreEqual(1, guild.HeroesCount);
		}
		[Test]
        public void RemoveHero_removeNonExistentHero_throwsException()
		{
			Guild guild = Guild.Create("my guild");
			Hero hero1 = Hero.Create("my hero 1");
			Hero hero2 = Hero.Create("my hero 2");
			guild.AddHero(hero1);
			var exc = Assert.Throws<InvalidOperationException>(
				() => guild.RemoveHero(hero2)
			);
			Assert.IsTrue(
				exc.Message.ToLower().Contains(
					"attempt to delete non-existent hero"
				)
			);
		}
		[Test]
        public void RemoveHero_removeExistHero_getHeroesReturnOk()
		{
			Guild guild = Guild.Create("my guild");
			Hero hero1 = Hero.Create("my hero 1");
			Hero hero2 = Hero.Create("my hero 2");
			guild.AddHero(hero1);
			guild.AddHero(hero2);
			guild.RemoveHero(hero1);

			var heroes = guild.Heroes;
			Assert.IsFalse(heroes.Contains(hero1));
            Assert.IsTrue(heroes.Contains(hero2));
			Assert.AreEqual(1, heroes.Count);
        }

        [Test]
        public void GetHeroesDailyPayment_creatNewGuild_equalsZero()
		{
			Guild guild = Guild.Create("my guild");
			Assert.AreEqual(0, guild.HeroesDailyPayment);
		}
		[Test]
        public void GetHeroesDailyPayment_twoHeroes_returnCorrect()
		{
			Guild guild = Guild.Create("my guild");
			Hero hero1 = Hero.Create("my hero 1");
			Hero hero2 = Hero.Create("my hero 2");
			hero1.DailyFee = 10;
			hero2.DailyFee = 20;
			guild.AddHero(hero1);
			guild.AddHero(hero2);

			Assert.AreEqual(30, guild.HeroesDailyPayment);
		}
	}
}
